packages/editor-ui/src/push-connection/useWebSocketClient.ts

```diff
@@ -0,0 +1,113 @@
+import { useReconnectTimer } from '@/push-connection/useReconnectTimer';
+import { ref } from 'vue';
+
+export type UseWebSocketClientOptions<T> = {
+	url: string;
+	onMessage: (data: T) => void;
+};
+
+/** Defined here as not available in tests */
+export const WebSocketState = {
+	CONNECTING: 0,
+	OPEN: 1,
+	CLOSING: 2,
+	CLOSED: 3,
+};
+
+/**
+ * Creates a WebSocket connection to the server. Uses reconnection logic
+ * to reconnect if the connection is lost.
+ */
+export const useWebSocketClient = <T>(options: UseWebSocketClientOptions<T>) => {
+	const isConnected = ref(false);
+	const socket = ref<WebSocket | null>(null);
+
+	/**
+	 * Heartbeat timer to keep the connection alive. This is an additional
+	 * mechanism to the protocol level ping/pong mechanism the server sends.
+	 * This is used the ensure the client notices connection issues.
+	 */
+	const heartbeatTimer = ref<ReturnType<typeof setInterval> | null>(null);
+
+	const startHeartbeat = () => {
+		heartbeatTimer.value = setInterval(() => {
+			socket.value?.send(JSON.stringify({ type: 'heartbeat' }));
+		}, 30_000);
+	};
+
+	const stopHeartbeat = () => {
+		if (heartbeatTimer.value) {
+			clearInterval(heartbeatTimer.value);
+			heartbeatTimer.value = null;
+		}
+	};
```

Comment:
Can we extract this functionality into a `useHeartbeat` composable? Could be reused for other polling-type checks later.